beater/api/config/agent/handler.go

```diff
@@ -61,6 +63,119 @@ var (
 	rumAgents = []string{"rum-js", "js-base"}
 )
 
+// DirectConfigurationHandler returns a request.Handler for replying to agent
+// central configuration requests. The handler parses the incoming request and
+// responds with the config matching the service.name/service.environment pair.
+//
+// TODO: What if no matching service.name/service.environment pair is found?
+func DirectConfigurationHandler(serviceConfigs []config.ServiceConfig, config *config.AgentConfig, defaultServiceEnvironment string) request.Handler {
+	cacheControl := fmt.Sprintf("max-age=%v, must-revalidate", config.Cache.Expiration.Seconds())
+
+	return func(c *request.Context) {
+		// error handling
+		c.Header().Set(headers.CacheControl, errCacheControl)
+
+		ok := c.RateLimiter == nil || c.RateLimiter.Allow()
+		if !ok {
+			c.Result.SetDefault(request.IDResponseErrorsRateLimit)
+			c.Write()
+			return
+		}
+
+		// We aren't forwarding the query any more, but the agents are
+		// sending the same request, so we can re-use `buildQuery` to
+		// extract the name/environment pair.
+		query, queryErr := buildQuery(c)
+		if queryErr != nil {
+			extractQueryError(c, queryErr, c.AuthResult.Authorized)
+			c.Write()
+			return
+		}
+		if query.Service.Environment == "" {
+			query.Service.Environment = defaultServiceEnvironment
+		}
+
+		conf, err := findMatchingServiceConfig(serviceConfigs, query)
+		if err != nil {
+			apm.CaptureError(c.Request.Context(), err).Send()
+			extractInternalError(c, err, c.AuthResult.Authorized)
+			c.Write()
+			return
+		}
+
+		// configuration successfully fetched
+		c.Header().Set(headers.CacheControl, cacheControl)
+		c.Header().Set(headers.Etag, fmt.Sprintf("\"%s\"", conf.Etag))
+		c.Header().Set(headers.AccessControlExposeHeaders, headers.Etag)
+
+		if conf.Etag == ifNoneMatch(c) {
+			c.Result.SetDefault(request.IDResponseValidNotModified)
+		} else {
+			c.Result.SetWithBody(request.IDResponseValidOK, conf.Settings)
+		}
+		c.Write()
+	}
+}
+
+func findMatchingServiceConfig(scs []config.ServiceConfig, query agentcfg.Query) (*agentcfg.Source, error) {
+	name, env := query.Service.Name, query.Service.Environment
+	var nameConf, envConf *config.ServiceConfig
+
+	for i, cfg := range scs {
+		if cfg.Service.Name == name {
+			nameConf = &scs[i]
+		}
+		if cfg.Service.Environment == env {
+			envConf = &scs[i]
+		}
+	}
+
+	if nameConf == nil {
+		return nil, fmt.Errorf("no config found for service.name=%s", name)
+	}
+
+	if envConf == nil {
+		// Nothing to merge, just return nameConf
+		return &agentcfg.Source{
+			Settings: nameConf.Config,
+			Etag:     nameConf.Etag,
+			Agent:    name,
+		}, nil
+	}
+	merged := merge(nameConf.Config, envConf.Config)
+	m, err := json.Marshal(merged)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
There should be no merging and the logic here should match what we do in Kibana, which is to find the first matching config block in descending order:

 - service name and environment both defined in criteria, both match queried/default service environment and name
 - only service name defined in criteria, matches the queried service name
 - only service environment defined in criteria, matches the queried/default service environment

Reply:
>only service environment defined in criteria, matches the queried/default service environment

Based on the pre-existing code, this isn't allowed:
https://github.com/elastic/apm-server/blob/master/beater/api/config/agent/handler.go#L170-L172

>service name and environment both defined in criteria, both match queried/default service environment and name

So in this situation, whichever is defined first under the `agent_config` (name or environment) is the config block returned?

Reply:
I think my confusion is stemming from `agent_config` example having `service.name` or `service.environment` in the issue, but the actual ui in kibana has you selecting both of these values.

Reply:
>> only service environment defined in criteria, matches the queried/default service environment
>
> Based on the pre-existing code, this isn't allowed:
https://github.com/elastic/apm-server/blob/master/beater/api/config/agent/handler.go#L170-L172

The agent must specify its service name. What I meant was that in the central config criteria the name is not specified.

> I think my confusion is stemming from agent_config example having service.name or service.environment in the issue, but the actual ui in kibana has you selecting both of these values.

In the UI there is a special value "All" for both name and environment, which is effectively the same as not setting it at all. So for the above scenario, in the existing UI that would be name set to "All" and environment set to something specific.